pennylane/beta/interfaces/tf.py

```diff
@@ -0,0 +1,187 @@
+                [tf.convert_to_tensor(p)]
+            ):
+                trainable_params.add(idx)
+
+        self.trainable_params = trainable_params
+
+    @staticmethod
+    def convert_to_numpy(tensors):
+        """Converts any TensorFlow tensors in a sequence to NumPy arrays.
+
+        Args:
+            tensors (Sequence[Any, tf.Variable, tf.Tensor]): input sequence
+
+        Returns:
+            list[Any, array]: list with all tensors converted to NumPy arrays
+        """
+        return [i.numpy() if isinstance(i, (tf.Variable, tf.Tensor)) else i for i in tensors]
+
+    @tf.custom_gradient
+    def _execute(self, params, **input_kwargs):
+        # unwrap free parameters
+        args = self.convert_to_numpy(params)
+
+        # unwrap constant parameters
+        all_params = self.get_parameters(trainable_only=False)
+        all_params_unwrapped = self.convert_to_numpy(all_params)
+
+        self.set_parameters(all_params_unwrapped, trainable_only=False)
+        res = self.execute_device(args, input_kwargs["device"])
+        self.set_parameters(all_params, trainable_only=False)
+
+        def grad(grad_output, **tfkwargs):
+            variables = tfkwargs.get("variables", None)
+
+            self.set_parameters(all_params_unwrapped, trainable_only=False)
+            jacobian = self.jacobian(input_kwargs["device"], params=args, **self.jacobian_options)
+            self.set_parameters(all_params, trainable_only=False)
+
+            jacobian = tf.constant(jacobian, dtype=self.dtype)
+
+            # Reshape gradient output array as a 2D row-vector.
+            # grad_output = tf.cast(grad_output, dtype=self.dtype)
+            grad_output_row = tf.transpose(tf.reshape(grad_output, [-1, 1]))
+
+            # Calculate the vector-Jacobian matrix product, and unstack the output.
+            grad_input = tf.matmul(grad_output_row, jacobian)
+            grad_input = tf.unstack(tf.reshape(grad_input, [-1]))
+
+            if variables is not None:
+                return grad_input, variables
+
+            return grad_input
+
+        if res.dtype == np.dtype("object"):
+            res = np.hstack(res)
+
+        return tf.convert_to_tensor(res, dtype=self.dtype), grad
+
+    @classmethod
+    def apply(cls, tape, dtype=tf.float64):
+        """Apply the TensorFlow interface to an existing tape in-place.
+
+        Args:
+            tape (.QuantumTape): a quantum tape to apply the TF interface to
+            dtype (tf.dtype): the dtype that the returned quantum tape should
+                output
+
+        **Example**
+
+        >>> with QuantumTape() as tape:
+        ...     qml.RX(0.5, wires=0)
+        ...     expval(qml.PauliZ(0))
+        >>> TFInterface.apply(tape)
+        >>> tape
+        <TFQuantumTape: wires=<Wires = [0]>, params=1>
+        """
+        tape_class = getattr(tape, "__bare__", tape.__class__)
+        tape.__bare__ = tape_class

```

Comment:
When would the `__bare__` attr be used?

Reply:
The only time it's used is on the line directly above 🙂 

This allows us to keep track of the original tape (sub)class. This way, if a user applies a TF interface to a ParamShiftTape, we can later apply a different interface and still end up with a wrapped ParamShiftTape.